### PR TITLE
DoAction - loadJSON and withInput actions

### PR DESCRIPTION
**Pull request by Paul-Doug**
Load JSON file on demand and apply pre-processed patches on user input.

### CODE REVIEW

**Comment by duboisp** [comment]
I think you already know, there is some linting issue that would need to be fixed: https://travis-ci.org/wet-boew/GCWeb/builds/595678809#L171

File: src/plugins/jsonmanager/jsonmanager-doc-en.hbs [hunk 1]
```diff
@@ -11,209 +11,182 @@
 ---
 <div class="wb-prettify all-pre"></div>
 <section>
```
Comment (duboisp) [inline src/plugins/jsonmanager/jsonmanager-doc-en.hbs]: Global comment for this file (jsonmanager-doc-en.hbs)

What has changed? this file look the same as the original.

File: src/plugins/doaction/doaction-doc-en.hbs [hunk 2]
```diff
@@ -78,43 +77,84 @@
...
+<dl class="dl-horizontal">
+<dt>{qval} string</dt>
+<dd>Query value in the url required to match a regular expression with <code>match</code> configuration. It can be specified by setting the parameter <code>qval</code>.</dd>
+<dt>{base} string</dt>
+<dd>Only for patches. Define the base string of a JSON pointer. Default value is <code>/</code>. It can be specified by setting the parameter <code>base</code>.</dd>
+</dl>
+<h3><code>match</code> and <code>default</code> configuration</h3>
+<p>If <code>match</code> is defined in your configuration, then the <code>default</code> configuration <strong>must</strong> be also defined.</p>
+<p>For security reason, any value retreived from the page URL always require to configure <code>match</code> that will contain a regular expresion. If there no match or the matching result is an empty string, the <code>default</code> configuration will be used as the value fallback.</p>
+<p>When you specify a regular expression, it is recommended to have it prefixed with "<code>^</code>" and suffixed with "<code>$</code>". Be careful to not use regular expression that could compromise the security.</p>
```
[the remark this answers is not in the document]
Comment (Paul-Doug) [reply src/plugins/doaction/doaction-doc-en.hbs]: Done

File: src/plugins/doaction/doaction-doc-en.hbs [hunk 3]
```diff
@@ -125,564 +165,583 @@
...
-				<li>addClass</li>
-				<li>removeClass</li>
-				<li>mapfilter</li>
-				<li>tocsv</li>
-			</ul>
-		</td>
-		<td><code>data-wb-doaction='{ "action": "tblfilter", "source": "#jQuerySelector" }'</code></td>
-		<td>jQuery selector and when the action is set to <code>tblfilter</code> the source should refer to a data table enhanced with <a href="https://wet-boew.github.io/wet-boew/demos/tables/tables-en.html">tables plugin</a>.</td>
-	</tr>
+<table id="actionConfig" class="table wb-tables" data-wb-tables='{ "ordering" : false, "paging": false }'>
```
[the remark this answers is not in the document]
Comment (Paul-Doug) [reply src/plugins/doaction/doaction-doc-en.hbs]: Done

File: src/plugins/doaction/doaction-doc-en.hbs [hunk 4]
```diff
@@ -125,564 +165,583 @@
...
-				<td><code>wb-ready.wb-doaction</code>
-					<ul>
-						<li>Plugin core</li>
-					</ul>
-				</td>
-				<td>Triggered automatically after an wb-doaction is initialized.</td>
-				<td>Used to identify when an wb-doaction has initialized (target of the event)
-					<pre><code>$( document ).on( "wb-ready.wb-doaction", ".wb-doaction", function( event ) {
+<h2>Events</h2>
+<table id="events" class="table wb-tables" data-wb-tables='{ "ordering" : false, "paging": false }'>
```
[the remark this answers is not in the document]
Comment (Paul-Doug) [reply src/plugins/doaction/doaction-doc-en.hbs]: Done

**Comment by duboisp** [comment]
Pre-approved at the technical review upon changes listed in my last comments

**Comment by Paul-Doug** [comment]
Made the requested changes. 
I can't see any issues in the Travis-ci report.

File: src/plugins/doaction/json-fr.hbs [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+		"patches" : {
+			"op": "move",
+			"path": "{base}",
+			"from": "{base}/{qval}"
+		}
+	}
+] }'>Appliquer un correctif basé sur l'entrée d'utilisateur</button>
+</div>
+<p id="jmexample3" data-wb-jsonmanager='{ "url": "../urlmapping/demo/data-en.json", "name": "example4", "wait": true }'>Données corrigées selon la valeur de l'entrée: <span data-json-replace="#[example4]/city">Rien n'est encore chargé</span></p>
```
Comment (GormFrank) [inline src/plugins/doaction/json-fr.hbs]: data-fr.json

File: src/plugins/doaction/json-fr.hbs [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+		&quot;action&quot;: &quot;patch&quot;, &quot;source&quot;: &quot;#jmexample3&quot;, &quot;base&quot;: &quot;/city&quot;, &quot;match&quot;: &quot;^[a-z][a-z][a-z]$&quot;, &quot;default&quot;: &quot;vdm&quot;,
+
+		&quot;patches&quot; : {
+			&quot;op&quot;: &quot;move&quot;,
+			&quot;path&quot;: &quot;{base}&quot;,
+			&quot;from&quot;: &quot;{base}/{qval}&quot;
+		}
+	}
+] }'&gt;Appliquer un correctif basé sur l'entrée d'utilisateur&lt;/button&gt;
+&lt;p id=&quot;jmexample3&quot; data-wb-jsonmanager='{ &quot;url&quot;: &quot;../urlmapping/demo/data-en.json&quot;, &quot;name&quot;: &quot;example4&quot;, &quot;wait&quot;: true }'&gt;Données corrigées selon la valeur de l'entrée: &lt;span data-json-replace=&quot;#[example4]/city&quot;&gt;Rien n'est encore chargé&lt;/span&gt;&lt;/p&gt;</code></pre>
```
Comment (GormFrank) [inline src/plugins/doaction/json-fr.hbs]: data-fr.json

File: src/plugins/doaction/json-fr.hbs [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+<h3>Fichier <code>demo/data-en.json</code></h3>
```
Comment (GormFrank) [inline src/plugins/doaction/json-fr.hbs]: data-fr.json

File: src/plugins/doaction/json-fr.hbs [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+"city": {
+"gat": "Gatineau",
+"ott": "Ottawa",
+"vdm": "Val-des-monts"
+},
+"fruit": "Perry",
+"country": {
+"name": "Canada",
+"province": {
+"qc": "Quebec",
```
Comment (GormFrank) [inline src/plugins/doaction/json-fr.hbs]: Québec

File: src/plugins/doaction/demo/data-en.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  "fruit": "Perry",
+  "country": {
+    "name": "Canada",
+    "province": {
+      "qc": "Quebec",
+      "on": "Ontario",
+      "ma": "Manitoba"
+    }
+  }
+}
```
Comment (GormFrank) [inline src/plugins/doaction/demo/data-en.json]: Needs the same file in French (with Québec) named data-fr.json

File: src/plugins/doaction/demo/doaction-date-fr.json [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{
+"test" : "test found!",
```
Comment (GormFrank) [inline src/plugins/doaction/demo/doaction-date-fr.json]: Test trouvé!

File: src/plugins/doaction/demo/doaction-date-fr.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+"express" : {
+"bad-date" : "Date not found",
```
Comment (GormFrank) [inline src/plugins/doaction/demo/doaction-date-fr.json]: Date introuvable

File: src/plugins/doaction/demo/doaction-date-fr.json [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+"2019-06-10" : "2019-06-13",
+"2019-06-11" : "2019-06-14",
+"2019-06-12" : "2019-06-15"
+},
+"standard" : {
+"bad-date" : "Date not found",
```
Comment (GormFrank) [inline src/plugins/doaction/demo/doaction-date-fr.json]: Date introuvable

File: src/plugins/doaction/demo/file-4.json [hunk 13]
```diff
@@ -0,0 +1,3 @@
+{
+  "data": "&From JSON file 4"
+}
```
Comment (GormFrank) [inline src/plugins/doaction/demo/file-4.json]: Needs French equivalent (file-4-fr.json)

`{
  "data": "&Du fichier JSON 4"
}`

... and is it `*` or `&`?

File: src/plugins/doaction/demo/file-3.json [hunk 14]
```diff
@@ -0,0 +1,3 @@
+{
+  "data": "-From JSON file 3"
+}
```
Comment (GormFrank) [inline src/plugins/doaction/demo/file-3.json]: Needs French equivalent (file-3-fr.json)

`{
  "data": "-Du fichier JSON 3"
}`

File: src/plugins/doaction/demo/file-2.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+  "data": "+From JSON file 2"
+}
```
Comment (GormFrank) [inline src/plugins/doaction/demo/file-2.json]: Needs French equivalent (file-2-fr.json)

`{
  "data": "+Du fichier JSON 2"
}`

File: src/plugins/doaction/demo/file-1.json [hunk 16]
```diff
@@ -0,0 +1,3 @@
+{
+  "data": "From JSON file 1"
+}
```
Comment (GormFrank) [inline src/plugins/doaction/demo/file-1.json]: Needs French equivalent (file-1-fr.json)

`{
  "data": "Du fichier JSON 1"
}`

File: src/plugins/doaction/json-fr.hbs [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+
+<p>Cette application est utilisé avec JSON-Manager et Data-JSON. La documentation sur l'application du correctif est documentée dans le gestionnaire JSON et dans le correctif de  mappage d'URL.</p>
+<h2>Télécharger un fichier JSON sur demande</h2>
+<p>Télécharger les fichiers</p>
+
+<ul>
+<li><button type="button" data-wb-doaction='{ "action": "loadJSON", "source": "#testLoader", "url": "demo/file-1.json" }'>JSON fichier 1</button></li>
+<li><button type="button" data-wb-doaction='{ "action": "loadJSON", "source": "#testLoader", "url": "demo/file-2.json" }'>JSON fichier 2</button></li>
+<li><button type="button" data-wb-doaction='{ "action": "loadJSON", "source": "#testLoader", "url": "demo/file-3.json" }'>JSON fichier 3</button></li>
+<li><button type="button" data-wb-doaction='{ "action": "loadJSON", "source": "#testLoader", "url": "demo/file-4.json" }'>JSON fichier 4</button></li>
```
Comment (GormFrank) [inline src/plugins/doaction/json-fr.hbs]: file-X-fr.json for the 4 of them

File: src/plugins/doaction/json-fr.hbs [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+<h3>Fichier <code>demo/file-2.json</code></h3>
+<pre><code>{"data": "+From JSON file 2"}</code></pre>
+<h3>Fichier <code>demo/file-3.json</code></h3>
+<pre><code>{
+"data": "-From JSON file 3"
+}</code></pre>
+<h3>Fichier <code>demo/file-4.json</code></h3>
+<pre><code>{
+"data": "*From JSON file 4"
+}</code></pre>
```
Comment (GormFrank) [inline src/plugins/doaction/json-fr.hbs]: file-X-fr.json for the 4 of them with their text content from the new files

File: src/plugins/doaction/demo/file-4.json (hunk 13 above)
Comment (duboisp) [reply src/plugins/doaction/demo/file-4.json]: Use & to be constant. But that was only to differentiate the data inside those multiple files

**Pull request merged**